pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+
+	type FungiblesAssetIdOf<T> =
+		<<T as Config>::Fungibles as InspectFungibles<<T as frame_system::Config>::AccountId>>::AssetId;
+
+	type CollateralAssetIdOf<T> =
+		<<T as Config>::CollateralCurrency as InspectFungibles<<T as frame_system::Config>::AccountId>>::AssetId;
+
+	type BoundedCurrencyVec<T> = BoundedVec<FungiblesAssetIdOf<T>, <T as Config>::MaxCurrencies>;
+
+	pub(crate) type CurrencyNameOf<T> = BoundedVec<u8, <T as Config>::MaxStringLength>;
+
+	pub(crate) type CurrencySymbolOf<T> = BoundedVec<u8, <T as Config>::MaxStringLength>;
+
+	pub(crate) type CurveParameterTypeOf<T> = <T as Config>::CurveParameterType;
+
+	// TODO: change CurveParameterTypeOf.
+	pub(crate) type PoolDetailsOf<T> =
+		PoolDetails<<T as frame_system::Config>::AccountId, CurveParameterTypeOf<T>, BoundedCurrencyVec<T>>;
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The currency used for storage deposits.
+		type DepositCurrency: MutateHold<Self::AccountId, Reason = Self::RuntimeHoldReason>;
+		/// The currency used as collateral for minting bonded tokens.
+		type CollateralCurrency: MutateFungibles<Self::AccountId>
+			+ AccountTouch<CollateralAssetIdOf<Self>, Self::AccountId>
+			+ FungiblesMetadata<Self::AccountId>;
+		/// Implementation of creating and managing new fungibles
+		type Fungibles: CreateFungibles<Self::AccountId, AssetId = Self::AssetId>
+			+ DestroyFungibles<Self::AccountId>
+			+ FungiblesMetadata<Self::AccountId>
+			+ FungiblesInspect<Self::AccountId>
+			+ MutateFungibles<Self::AccountId, Balance = CollateralCurrencyBalanceOf<Self>>;
+		/// The maximum number of currencies allowed for a single pool.
+		#[pallet::constant]
+		type MaxCurrencies: Get<u32>;
+		/// The deposit required for each bonded currency.
+
+		#[pallet::constant]
+		type MaxStringLength: Get<u32>;
+
+		/// The deposit required for each bonded currency.
+		#[pallet::constant]
+		type DepositPerCurrency: Get<DepositCurrencyBalanceOf<Self>>;
+
+		/// The base deposit required to create a new pool, primarily to cover the ED of the pool account.
+		#[pallet::constant]
+		type BaseDeposit: Get<DepositCurrencyBalanceOf<Self>>;
+
+		/// The asset id of the collateral currency.
+		type CollateralAssetId: Get<CollateralAssetIdOf<Self>>;
+		/// Who can create new bonded currency pools.
+		type PoolCreateOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Self::AccountId>;
+		/// The type used for pool ids
+		type PoolId: Parameter + MaxEncodedLen + From<[u8; 32]> + Into<Self::AccountId>;
+
+		/// The type used for asset ids. This is the type of the bonded currencies.
+		type AssetId: Parameter + Member + FullCodec + TypeInfo + MaxEncodedLen + Saturating + One + Default;
```

Comment:
Isn't this already part of the `Inspect` trait?

Reply:
yeah should be

Reply:
this is our way to express the trait bounds on the Fungible's asset id type though - if there is a better way, let us know!

Reply:
You can add bounds on the associated type in the `impl` blocks, as done for instance [here](https://github.com/KILTprotocol/kilt-node/blob/4148afb2926a0567432e0b0c6546de98fa30a759/pallets/pallet-asset-switch/src/lib.rs#L152).

Reply:
That's not relevant though if we move to an next asset id hook, which we should try to do first.